utils/logalloc: introduce segment_store_backend

denesb opened the pull request:
We want to make it possible to switch between segment store implementations run-time but without making all members of the segment store virtual. The role of the segment stores is to allocate/deallocate and assign ids to segments. The seastar allocator based one does this based on a memory-layout description obtained from the seastar allocator, while the standard allocator based one maintains a vector of segments to provide the same services. We observe that it is possible to standardize on the former if we also allocate a large contiguous memory area for segments when using the standard allocator. We achieve that by using mmap() and abstract the way segments are allocated/deallocated with a virtualized segment store backend. We can then reuse the accounting code of the seastar allocator based segment store largely untouched, with a virtual call needed (to the store backend) only when actually allocating/deallocating segments. With this we can have a single segment store implementation, with the correct backend selected using the preprocessor. This lays the groundwork for being able to instantiate or switch to a different backend.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1973/

denesb commented:
New in v2:
* Fix scylla-gdb tests.

tgrabiec commented:
What's the motivation for switching at trun-time?

denesb commented:
> What's the motivation for switching at run-time?

Tools. They are part of a release-mode production executable but they want to run with the standard allocator backend to avoid having having to commit to a certain memory amount, like scylla does normally. If they commit to too little, it might not be enough in some cases, if they commit to too much, it might clash with a running scylla instance.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1976/

denesb commented:
New in v3:
* Fix debug tests

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1991/

avikivity commented:
An alternative for tools is to run with index caching disabled (and maybe with the kernel page cache enabled).

avikivity commented:
(Although we might find ourselves eventually in a situation where cache is absolutely required, after we make things even more complicated)

avikivity commented on utils/logalloc.cc:
```diff
@@ -788,28 +788,36 @@ class mmap_memory_segment_store_backend : public segment_store_backend {
+#ifndef SEASTAR_DEFAULT_ALLOCATOR
+        return std::make_unique<seastar_memory_segment_store_backend>();
+#else
+        // We want to simulate 3GB mem per shard (1 << 30).
```
3GB = 3 << 30

avikivity replied:
and why 3GB?

avikivity replied:
I guess just a typo. For tools 1GB is likely too much, but it's not a real problem.

denesb replied:
I copied the comment and calculation from the existing std memory backed segment store: https://github.com/scylladb/scylladb/blob/fe401f14de638668ce539f589da9e6b8bb71814b/utils/logalloc.cc#L731

denesb replied:
Actually, yes its 1GB.

denesb replied:
Note that we just `mmap()` this memory but don't `mlock` it, so if no one uses it, the kernel doesn't have to materialize the backing. I will still use less for tools, will leave 1GB as the default (for tests).

denesb replied:
Although lsa will allocate all segments it can on startup, but doesn't write to them. It is a good question whether this will trigger assigning physical memory to these pages.

avikivity replied:
Why do we use mmap? This will reduce the ability of asan to track it.

avikivity replied:
IIRC the code has support for poison/unpoison already.

denesb commented:
This PR makes debug tests very sensitive on non-lsa pointers freed with default allocator and vice-versa, as the two (LSA and default allocator) have non-overlapping memory pools.

denesb commented:
New in v4:
* Fix typo in comment saying how much memory we allocate for debug.
* Fix calculation of `_segments_base`, use `align_up()` instead of `align_down()` so aligned address is not before the mapped memory.
* Fix prepending of new `free_segment` to free-chain: previously prepending resulted in overriding previous list head, instead of prepending the new element before it.
* Explicitly handle the case of an index-lookup for a segment not owned by the store.
* Make `test_compaction_with_multiple_regions` (`logalloc_test.cc`) exception safe.

denesb commented:
New in v4.1:
* Simplify prepend to freelist (no special case).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2033/

avikivity commented on utils/logalloc.cc:
```diff
@@ -817,16 +833,20 @@ class segment_store {
+            return _backend->free_memory() >= non_lsa_reserve + segment::size;
+        } else {
+            return _backend->free_memory() >= segment::size;
+        }
```
Should we push this logic to the backend instead of switching on some boolean?

I don't understand this completely.

It's better to expose a constant as a non-virtual function (as a micro optimization)

avikivity commented on utils/logalloc.cc:
```diff
@@ -846,80 +846,6 @@ class segment_store {
-        return i != _segments.end();
-    }
-};
-#endif
```
This style of building a replacement in a different place and then deleting the original in a different patch results in small patches, but they're hard to review. How can I tell some detail wasn't lost?

denesb replied:
One of the goals of this series was to discard the vector-based `segment_store` and keep the `memory_layout`-based one only, with the differences in how that memory is acquired and managed moved into the new `segment_store_backend`.
The `memory_layout` based store implementation is modified in-place. We don't want to keep anything from this one.

avikivity commented on utils/logalloc.cc:
```diff
@@ -693,6 +693,99 @@ struct segment_descriptor : public log_heap_hook<segment_descriptor_hist_options
+    }
+};
+
+class mmap_memory_segment_store_backend : public segment_store_backend {
```
I don't understand why we need a mmap backend. The debug allocator (below) uses aligned_alloc.

As far as I can tell, the critical different is using memory layout for the  segment address -> descriptor translation, vs. using a map.

denesb replied:
I will change this to go through the allocator. Indeed there is not much point in using mmap.

denesb replied:
The problem with the current state of art is that the the two stores use completely different method to map segment address -> descriptor translation. If I wanted to fit both into a common interface I'd have to virtualize the segment address to descriptor mapping, which is quite commonly used. To avoid that I standardize  on the memory layout based mapping and just virtualize how the backing memory is obtained and managed, as this is used much more rarely.

avikivity replied:
Ah, so mmap is used obtain all the segments, not an individual segment?

I'm confused. We already have two different segment -> descriptor mappings (for release and debug mode) and two different ways to obtain the memory. You're adding a third way to obtain memory, but using the division method for segment->descriptor?

It's probably the right thing if translations are frequent.

denesb replied:
> Ah, so mmap is used obtain all the segments, not an individual segment?

Yes. I wanted to emulate the seastar allocator which owns a large contiguous area of memory, allocating some of it to segments. This allows for simple segment -> descriptor mapping, via calculating the segment's offset in the layout and using that as an index. I wanted to emulate this in debug mode, obtaining a large continuous area of memory, dedicated to segments, so we can use a single mapping logic across release/debug modes.
I used mmap to obtain this memory, but getting it from the standard allocator works just as well.

After this patch we have two methods for obtaining the memory area that will be used to back segments, and a single method for mapping segments to descriptors.

avikivity replied:
It's not good for debug mode, since if we overrun a segment, we hit another segment without noticing. One of asan's benefits is to prevent overruns and we lose it with the release mode segment allocation.

denesb replied:
We can leave gaps between, although this will increase waste (gaps would have to be segment sized, as segments are aligned to `segment::size`). I wonder if poisoning forces the kernel to back poisoned pages with real memory. Otherwise, the waste might remain theoretical only.

Or we could just keep the current debug-mode segment store for debug-only and use the new one for tools only. I'd like to avoid that if possible.

avikivity replied:
No, poisoning wastes the memory. asan uses separate mmaps to large trap out-of-bounds violations and poisoning to trap small violations and uninitialized variable use.

Why not keep the debug-mode segment store? It works, and messing with it is liable to reduce our ability to catch bugs (since we don't test it).

denesb replied:
Ok, I will keep the current debug store.

denesb commented:
New in v5:
* Use `aligned_alloc()` instead of `mmap()` to obtain memory area for standard-allocator based backend. This backend is renamed to `standard_memory_segment_store_backend`.
* Added comments to each backend explaining how they work.
* Moved logic of `can_allocate_more_segments()` into backend and changed the name of the overlap flag to something hopefully more descriptive.

denesb commented:
Closing this because due to scope change, the new segment backend would go in unused, prone to rot. Will submit a separate PR with the intended usage also present.

denesb commented:
Supplanted by https://github.com/scylladb/scylladb/pull/11510
